uploader/uploader.go

```diff
@@ -1,149 +1,127 @@
 package uploader
 
 import (
-	"fmt"
+	"context"
+	"io"
 	"io/ioutil"
 	"os"
 	"path"
+	"sync"
 
-	"io"
-
-	"context"
+	"github.com/Netflix/titus-executor/api/netflix/titus"
+	"github.com/hashicorp/go-multierror"
 
 	"github.com/Netflix/metrics-client-go/metrics"
 	"github.com/Netflix/titus-executor/config"
 	log "github.com/sirupsen/logrus"
 )
 
-func collectCopyUploaders(cfg *config.Config) []Uploader {
-	copyUploaders := []Uploader{}
-	for _, uploaderArg := range cfg.CopyUploaders {
-		copyUploaders = append(copyUploaders, NewCopyUploader(uploaderArg))
-	}
+const (
+	s3BucketNameParam = "titusParameter.agent.log.s3BucketName"
+	s3PathPrefixParam = "titusParameter.agent.log.s3PathPrefix"
+)
 
-	return copyUploaders
+// Wraps the backends used to do log uploading.
+type Uploader struct {
+	backend Backend
 }
 
-func collectS3Uploaders(cfg *config.Config, m metrics.Reporter) []Uploader {
-	copyUploaders := []Uploader{}
-	for _, uploaderArg := range cfg.S3Uploaders {
-		copyUploaders = append(copyUploaders, NewS3Uploader(m, log.StandardLogger(), uploaderArg))
-	}
-
-	return copyUploaders
-}
+// The upload always prefers s3, then copy, and will use a black hole sink if nothing else is configured. The first
+// s3 location or copy destination specified is used. Rest are ignored.
+func NewUploader(config *config.Config, titusInfo *titus.ContainerInfo, taskID string, m metrics.Reporter) (*Uploader, error) {
+	bucketName, pathPrefix := "", ""
 
-func collectNoopUploaders(cfg *config.Config) []Uploader {
-	copyUploaders := []Uploader{}
-	for range cfg.NoopUploaders {
-		copyUploaders = append(copyUploaders, NewNoopUploader())
+	if len(config.S3Uploaders) > 0 {
+		bucketName = config.S3Uploaders[0]
 	}
 
-	return copyUploaders
-}
+	if param, ok := titusInfo.GetPassthroughAttributes()[s3BucketNameParam]; ok {
+		bucketName = param
```

Comment:
What's the behavior if we can't write to this S3 bucket?  Do we fall back to the default?

Reply:
We drop the logs. The concern here is we write some sensitive information to a default bucket which does not have the right permissions.